Add device actions

Steve-Mcl opened the pull request:
**Ready to review, but please do not merge until both this and https://github.com/FlowFuse/device-agent/pull/239 are approved**

## Description

* Adds suspend/start/restart support for devices backend API in `forge/routes/api/deviceActions.js`
   * POST `/api/v1/devices/:deviceId/actions/suspend`
   * POST `/api/v1/devices/:deviceId/actions/start`
   * POST `/api/v1/devices/:deviceId/actions/restart`
* Adds suspend/start/restart support for devices frontend API in `frontend/src/api/devices.js`
   * suspendDevice
   * startDevice
   * restartDevice
* Adds UI for device actions (Actions Dropdown)
* Adds audit logging for actions
* Adds unit tests for each action
* Informs user if agent version reported is less than `V2.3` (see issue comments)

### UI Changes (discussed with @joepavitt )
* The new Actions Dropdown was placed to the far right (aligns with instance)
* The device developer mode toggle was moved to the left of the "Device Editor" (since the two are intrinsically linked)
* small gap-y-2 added to the button group to improve appearance on small screen (when buttons wrap to left)
* Device mutation handler added for transitional states (same as instance)
  * This includes polling for device state when in a transitional state


### Tests added:
```
Audit Log > Device
    ✔ Provides a logger for device started action
    ✔ Provides a logger for device start failed action
    ✔ Provides a logger for device restarted action
    ✔ Provides a logger for device restart failed action
    ✔ Provides a logger for device suspended action
    ✔ Provides a logger for device suspend failed action

  Device API
    Device Actions
      with MQTT comms
        ✔ team owner can trigger device restart action
        ✔ team owner can trigger device start action
        ✔ team owner can trigger device suspend action
        ✔ restart action returns 400 restart_failed when agent responds with success:false
        ✔ start action returns 400 start_failed when agent responds with success:false
        ✔ suspend action returns 400 suspend_failed when agent responds with success:false
        ✔ restart action returns 400 device_suspended when agent is suspended
        ✔ suspend action returns 400 device_suspended when agent is suspended
        ✔ team member cannot trigger restart action
        ✔ team member cannot trigger start action
        ✔ team member cannot trigger suspend action
        ✔ offline device times out for restart action (38ms)
        ✔ offline device times out for start action (44ms)
        ✔ offline device times out for suspend action (42ms)
      without MQTT comms
        ✔ restart results in error 400
        ✔ start results in error 400
        ✔ suspend results in error 400
```

### ~~TODO~~
* Docs
   * DONE
* Handle _No device comms/MQTT_. Currently, the "Actions Dropdown" is available in UI regardless, but will error (400) if comms not available.
   * DONE: see known issues

### Known issues
* Frontend does not know if comms on the platform or on the device have been configured. Actions will be available but will time out.  This decision was made following discussion in #3292 



## Related Issue(s)

#3292

## Checklist

<!-- https://flowfuse.com/handbook/development/#defining-done -->

 - [x] I have read the [contribution guidelines](https://github.com/FlowFuse/flowfuse/blob/main/CONTRIBUTING.md)
 - [x] Suitable unit/system level tests have been added and they pass <!-- If not adding test coverage, please clarify why not? -->
 - [x] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [x] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `FlowFuse/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `FlowFuse/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Backport needed? -> add the `backport` label
 - [ ] Includes a DB migration? -> add the `area:migration` label

knolleary commented on docs/user/concepts.md:
```diff
@@ -159,6 +160,22 @@ To further simplify device registration, Provisioning Tokens can be created to a
 devices to automatically connect to a team without having to manually register them first.
 The token can also be configured to assign a device directly to a Node-RED instance within the team.
 
+### Device Actions
```
Not 100% sure if this warrants being a top-level concept. It feels more like a feature of devices that should be in the device specific docs.

Steve-Mcl replied:
Unfortunately, there are no /user/device specific docs and /device-agent docs are about the agent and its installation/registration. 

possible options:
1. remove this all together
   1. viable - there is no counterpart "instance actions" in the docs (only a smattering of "... requires a restart of the Instance")
2. keep it and add a follow up task to generate full user/device docs and move it
3. generate full user/device docs as part of this PR
4. other?

^ @knolleary thoughts?

knolleary replied:
I would pull it out of this PR. The docs need more work than we should try to contain in this one PR. In the short term, this feature is reasonably self-explanatory in the UI so doesn't need lots of accompany documentation.

Steve-Mcl replied:
resolved in [8e83791](https://github.com/FlowFuse/flowfuse/pull/3548/commits/8e837918fab03c127f20f5b7d771edcacdcd08b6), task here: https://github.com/FlowFuse/flowfuse/issues/3566

knolleary commented on forge/db/controllers/Device.js:
```diff
@@ -1,7 +1,45 @@
+ * is no need to store that in the database. But we do need to know it so the
+ * information can be returned on the API.
+ */
+const inflightState = { }
```
We cannot have inflight state - it will prevent us from being able to horizontally scale the app. Each API request could hit a different instance of the platform.

I haven't looked to see how exactly this state is used in this PR but will need addressing.

Steve-Mcl replied:
hmmmm. good point and no simple solution. 

Will need to study.

Steve-Mcl replied:
@knolleary I have removed the in-memory inflight state and reverted to updating the device.  tests updated to reflect this.

the UI already optimistically sets state so when the real status comes in from the API, it will align with the real DB value.

Operationally tested and works consistently.

This approach should be suitable for scaling.

Update incoming.

Steve-Mcl replied:
Update here: [remove inflight device status](https://github.com/FlowFuse/flowfuse/pull/3548/commits/ae52707d70d7a1d71c1921ad65d6f4abfdf02dae)

knolleary commented on forge/lib/permissions.js:
```diff
@@ -71,6 +71,7 @@ const Permissions = {
     'device:delete': { description: 'Delete a Device', role: Roles.Owner },
     'device:edit': { description: 'Edit a Device', role: Roles.Owner },
     'device:edit-env': { description: 'Edit Device Environment Variables', role: Roles.Member },
+    'device:change-status': { description: 'Start/Stop Project', role: Roles.Owner },
```
Incorrect description - not a project

knolleary reviewed (changes requested):
Comments from first pass review. Will get it running tomorrow to try it out and look at frontend changes.

Steve-Mcl commented on forge/lib/permissions.js:
```diff
@@ -71,6 +71,7 @@ const Permissions = {
     'device:delete': { description: 'Delete a Device', role: Roles.Owner },
     'device:edit': { description: 'Edit a Device', role: Roles.Owner },
     'device:edit-env': { description: 'Edit Device Environment Variables', role: Roles.Member },
+    'device:change-status': { description: 'Start/Stop Project', role: Roles.Owner },
```
```suggestion
    'device:change-status': { description: 'Start/Stop a Device', role: Roles.Owner },
```

Steve-Mcl replied:
@knolleary ^ better?

Steve-Mcl commented on docs/user/concepts.md:
```diff
@@ -159,6 +160,22 @@ To further simplify device registration, Provisioning Tokens can be created to a
 devices to automatically connect to a team without having to manually register them first.
 The token can also be configured to assign a device directly to a Node-RED instance within the team.
 
+### Device Actions
+
+**Introduced in FlowFuse 2.2**
+
+Devices have an action dropdown menu that can request the device agent to suspend, start and restart the
+Node-RED instance running on the device.
+
+This is useful when you want to temporarily stop Node-RED, restart it to pick up updates or to aid diagnosing issues.
+
+While a device is suspended, the Device Agent will continue to run and report device status
+to the platform, but your Node-RED flows will not stopped. If the Device Agent is restarted,
+Node-RED will be continue to be suspended. In order to exit suspended mode, start action must be requested.
+
+NOTE: Device Actions require FlowFuse to be configured with a broker. 
+Please contact your FlowFuse administrator for more information or see the [broker configuration](../install/configuration.md#mqtt-broker-configuration) documentation.
+
```
```suggestion
```

remove docs, feature is self explaining - for later follow up task

Steve-Mcl replied:
Applying this patch - overreaching issue addressed in https://github.com/FlowFuse/flowfuse/issues/3566

joepavitt commented:
@Steve-Mcl just checking on this, how we looking for tomorrow? Just the test issues that are the blocker atm?

Steve-Mcl commented:
@joepavitt tests are fixed. I've already re-requested review. Good to go afaik 🤷‍♂️

It's counterpart was reviewed by @hardillb but as requested in the PR, he has not merged it in case structural changes were thrown up in either review (as the 2 are linked)

knolleary commented:
Not liking how the header now wraps the content:

<img width="1387" alt="image" src="https://github.com/FlowFuse/flowfuse/assets/51083/95ef83ba-c568-4e0c-aa42-398c7396d3f5">

This is a 1300px wide screen - which isn't unreasonable small and typically about how wide I have my browser

Steve-Mcl commented:
> Not liking how the header now wraps the content:
> 
> <img alt="image" width="1387" src="https://private-user-images.githubusercontent.com/51083/312532869-95ef83ba-c568-4e0c-aa42-398c7396d3f5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAzNDU5NzMsIm5iZiI6MTcxMDM0NTY3MywicGF0aCI6Ii81MTA4My8zMTI1MzI4NjktOTVlZjgzYmEtYzU2OC00ZTBjLWFhNDItMzk4YzczOTZkM2Y1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzEzVDE2MDExM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMxZWM5OTRkYjczZGQ3NGJhZTdkZjllNTUwMTZmODcyNzczNzU1MmU1OTZiZWUxMTY1ODRmYjNkZGIxNzg5MTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.85A_iNVNMxdT0VFlb_1kf56QgzmmaWtNmWjalxLoMHo">
> This is a 1300px wide screen - which isn't unreasonable small and typically about how wide I have my browser

Agree.  We have an extra status pill and extra toggle button making this more frequent than instances.

joepavitt commented:
> This is a 1300px wide screen - which isn't unreasonable small and typically about how wide I have my browser

I'll have a play on the branch and see if I can come up with a better CSS layout

joepavitt commented:
Also this

## Before

<img width="687" alt="Screenshot 2024-03-13 at 17 33 19" src="https://github.com/FlowFuse/flowfuse/assets/99246719/ab1f92da-4064-4932-bc0c-e06f23a73326">

## After
<img width="683" alt="Screenshot 2024-03-13 at 17 33 04" src="https://github.com/FlowFuse/flowfuse/assets/99246719/d591fe1e-3037-4c9d-b586-eb093d3c7b2c">

joepavitt commented:
As discussed in #3292 - let's pull "Suspend" as an action for now due to ambiguity over billing consequences, and instead just include the "Restart" action with a follow up item

Steve-Mcl commented:
> As discussed in #3292 - let's pull "Suspend" as an action for now due to ambiguity over billing consequences, and instead just include the "Restart" action with a follow up item

I will hide these items on the menu and disable the APIs

Steve-Mcl commented:
* Removed actions from Action Dropdown
   * Commented out the API to prevent publishing API schema (and prevent endpoints being called)
   * Commented out the actions (see GIF)
* Removed `flex-1` from button group to minimise excessive padding causing early wrapping
   * Okayed with @joepavitt
   * see GIF below

![explorer_tjBjC4NX5k](https://github.com/FlowFuse/flowfuse/assets/44235289/155a3125-71f4-4e2a-9a58-1e23e38eb337)

Steve-Mcl commented:
And what did i forget to do? Update tests 🤦‍♂️

At least I remembered before I got the notification!

knolleary commented on docs/user/concepts.md:
```diff
@@ -22,6 +22,7 @@ them when _snapshots_ are created of the instance.
    - [Template](#template)
    - [Snapshot](#instance-snapshot)
  - [Device](#device)
+ - [Device Action](#device-actions)
```
```suggestion
```
No longer needed

knolleary commented:
Ok to merge once tests complete
